### ISSUE TITLE
Multi-branch support breaks concurrency scenarios (like monorepos)

### ISSUE THREAD

**Issue by yinzara** [semantic-release/semantic-release#1545]
## Current behavior

When running `semantic-release` concurrently on multiple projects within the same git repository when multiple "branches" are specified in the configuration, the executions fail before getting to the verify step of the plugins with errors like:
```
Error: Command failed with exit code 1: git fetch --tags --update-head-ok ssh://git@*********.org/myowner/myrepo.git +refs/heads/develop:refs/heads/develop
error: cannot lock ref 'refs/heads/develop': is at c4b425a6051fb1b2bc7cdded1265029c18850e4f but expected e1ca1dc0d76f5add27cf98d383ae23a3442fb5ee
```

When only a single entry in branches is specified, this functions normally with no errors.

This is caused by the `lib/branches/index.js` attempting to fetch all the branches named in the configuration from the repo repository.  `git fetch` (and other commands that write to the git repo) cannot be made concurrently with other requests that write to the repo.  When only a single branch is specified, there is no attempt to do this (hence why it works when only one branch is specified).

Before the multi-"branches" functionality was added, `semantic-release` worked great using plugins like `semantic-release-monorepo` even in large repositories by allowing concurrency (even though `semantic-release-monorepo` suggests you run with a concurrency of 1) and using special plugins that used external lock files to synchronize writes to the git repository.

Unfortunately this "work around" is impossible for a multiple release branch configuration because the loading of the branches is done before the plugin execution phase begins.

While I know there is a thread about how to enhance semantic-release to support monorepos, I felt this was a piece of low hanging fruit we could correct now to allow better support with current tools.

## Expected behavior

I believe there are a few possible solutions to this:
1. Introduce a lock file.  By introducing a shared lock file (and possibly a configuration entry of where it should be) semantic-release could synchronize the execution of git write operations outside of the current running process and prevent multiple threads from writing to the git repository at the same time. I believe this could be done safely and possibly without additional configuration though it might be wise to add it.

2. Add a configuration parameter that allows skipping any git write operations outside of the plugin executions.  Something like "skipFetchRemote" or "skipFetchBranches".  This would allow users of `semantic-release` to handle fetching manually before running semantic release and skip making semantic-release do it.  You could also just enhance the branch configuration to have a "skipFetch" config option as well that skipped fetching that specific branch. I don't love the idea of adding more configuration properties for special scenarios, but it's a simple and elegant solution. 

3. Do not fetch unless necessary.  i.e. change all calls to fetch to actually only scan for refs of the remote repo and after check if the local repo has those refs before fetching.  This would increase the number of remote git calls for all users of semantic-release for an edge case that isn't common (i.e. monorepos).   I also think there may still be concurrency problems.

4. Move the loading of the branches to become a plugin phase. This would mean moving the branch loading functionality to the @semantic-release/git plugin.  This is a pretty dramatic move but it provides for even more complex scenarios than this.

**Comment by IGZalbertoiglesias** [comment]
i am facing the same issue, I am working with monorepo and my pipelines files when i have to launch semantic release monorepo jobs in parallel.

`hint: Updates were rejected because the remote contains work that you do
 hint: not have locally. This is usually caused by another repository pushing
 hint: to the same ref. You may want to first integrate the remote changes
 hint: (e.g., 'git pull ...') before pushing again.`

**Comment by dhoulb** [comment]
If I understand the issue correctly it should be an easy fix: just change calls to `execa()` to `execa.sync()` in the codebase.

I hit concurrency issues like this in [multi-semantic-release](https://github.com/dhoulb/multi-semantic-release). The issues were resolved by hot patching `execa()` to `execa.sync()`

**Comment by yinzara** [comment]
I don't think it's quite that simple :) you have multiple processes all trying to concurrently execute git commands that modify the repository. Git does not support this. You must externally sync the commands via some algorithmic method (a few options outlined above).

**Comment by dhoulb** [comment]
@yinzara I’m pretty sure it is that simple! 

If you use `execa.sync()` instead of `execa()` to run Git commands, then you’ll never execute two at once because the program will wait for that command to complete before doing anything else. No two commands can be run concurrently.

Yes, you’d be locking up the thread and preventing other work from happening, but since semantic-release is primarily designed to release a single package I don’t think that’s a problem.

All plugins that use Git would need to make the same change to support multi-package release, but you’ll have that problem with any solution.

**Comment by yinzara** [comment]
When I run semantic-release, I'm running it in multiple processes in a monorepo using `lerna` and `semantic-release-monorepo`.   i.e. I am literally have multiple nodejs executable processes running.

`execa.sync` would only keep synchronization against the same process. I need synchronization across multiple node processes.

**Comment by dhoulb** [comment]
@yinzara Ah, gotcha! Hadn't considered that. I assume your project works if you use Lerna `--concurrency=1` right?

With the lockfile, you're suggesting a mechanism that would pause the entire `semantic-release` process until any previous `semantic-release` process had completed? If so, since you wouldn't get any performance benefit from concurrency or parallel mode, then the flag seems like a reasonable solution.

My current opinion is that semantic release needs a major internal rework to handle monorepos and concurrency natively. Any solution that wraps semantic release to work with monorepos or Lerna right now is going to be hacky and unstable and require significant parallel maintenance work.

**Comment by yinzara** [comment]
Yes that's right. If I run with "concurrency=1" everything works great.

And I think again you're going a bit too far. :-)

I'm not suggesting pausing the entire semantic release process. I'm only saying that when it performs operations that write to git (i.e. git fetch), it should use a lock file to prevent two threads from doing it at the exact same time.  That should only prevent two concurrent semantic release processes from writing to git at the same time, something that is a very small piece of what semantic release does at the beginning of its process cycle.

Synchronization during the plugin phases can always be done via configuration so there is only a need to make sure outside of plugins phases, that semantic release is thread safe. 

This was already true until the multi branch support got added.  When this was added, a bunch of "git fetch" operations now occur for branches that aren't the current branch before the plugin phases begin. This has broken concurrency scenarios.

I suggest adding a lock file to synchronize those fetches.  This would allow me to use unlimited concurrency.
